llvm/include/llvm/CodeGen/Analysis.h

```diff
@@ -62,16 +65,20 @@ inline unsigned ComputeLinearIndex(Type *Ty,
 /// If Offsets is non-null, it points to a vector to be filled in
 /// with the in-memory offsets of each of the individual values.
 ///
+/// If IRTypes is non-null, its pointee will contain the corresponding
+/// llvm::Types for each element in ValueVTs.
 void ComputeValueVTs(const TargetLowering &TLI, const DataLayout &DL, Type *Ty,
                      SmallVectorImpl<EVT> &ValueVTs,
                      SmallVectorImpl<EVT> *MemVTs,
                      SmallVectorImpl<TypeSize> *Offsets = nullptr,
-                     TypeSize StartingOffset = TypeSize::getZero());
+                     TypeSize StartingOffset = TypeSize::getZero(),
+                     SmallVectorImpl<Type *> *IRTypes = nullptr);
```

Comment:
NEED EXPLANATION

I don't understand this change.
Looking at how it is used in the ABI lowering, I'm surprised what existed was not enough.
My take is either you are holding LLVM wrong, or you have found a legitimate limitation.

In any case, this change will need some explaining/discussion when pushed to upstream.